From e81d6797a67e90f5402cbdea1ac31af66e2c62d5 Mon Sep 17 00:00:00 2001
From: Alistair Leslie-Hughes <leslie_alistair@hotmail.com>
Date: Mon, 22 Nov 2021 16:17:01 +1100
Subject: [PATCH 15/15] msdasql: HACK - IColumnsInfo GetColumnInfo for
 ICommandText

This needs to return columns.

This patch shows that ICommandPrepare_Prepare needs to be implement
by using SQLPrepare with a valid connection.  Then we can query the
columns without issues.  We should cache these values, as they aren't
going to change between GetColumnInfo(Command), and GetColumnInfo(Rowset).
---
 dlls/msdasql/session.c        | 15 +++++-
 dlls/msdasql/tests/provider.c | 96 +++++++++++++++++++++++------------
 2 files changed, 77 insertions(+), 34 deletions(-)

diff --git a/dlls/msdasql/session.c b/dlls/msdasql/session.c
index fd128ec69ee..3bc87d5d14d 100644
--- a/dlls/msdasql/session.c
+++ b/dlls/msdasql/session.c
@@ -1703,7 +1703,20 @@ static HRESULT WINAPI colsinfo_GetColumnInfo(IColumnsInfo *iface, DBORDINAL *col
 {
     struct command *command = impl_from_IColumnsInfo( iface );
     FIXME("%p, %p, %p, %p\n", command, columns, colinfo, stringsbuffer);
-    return E_NOTIMPL;
+
+    if (!command->query)
+        return DB_E_NOCOMMAND;
+
+    if (columns)
+        *columns = 0;
+
+    if (colinfo)
+        *colinfo = NULL;
+
+    if (stringsbuffer)
+        *stringsbuffer = NULL;
+
+    return DB_E_NOTPREPARED;
 }
 
 static HRESULT WINAPI colsinfo_MapColumnIDs(IColumnsInfo *iface, DBORDINAL column_ids,
diff --git a/dlls/msdasql/tests/provider.c b/dlls/msdasql/tests/provider.c
index 150aa4c6103..f4606d2ce9b 100644
--- a/dlls/msdasql/tests/provider.c
+++ b/dlls/msdasql/tests/provider.c
@@ -44,6 +44,37 @@ static char mdbpath[MAX_PATH];
 
 static const VARTYPE intptr_vartype = (sizeof(void *) == 8 ? VT_I8 : VT_I4);
 
+static char *dump_flags(DWORD flags)
+{
+    static char buffer[2048];
+
+    memset(buffer, 0, sizeof(buffer));
+    if(flags & DBCOLUMNFLAGS_ISBOOKMARK)
+        strcat(buffer, "DBCOLUMNFLAGS_ISBOOKMARK | ");
+    if(flags & DBCOLUMNFLAGS_MAYDEFER)
+        strcat(buffer, "DBCOLUMNFLAGS_MAYDEFER | ");
+    if(flags & DBCOLUMNFLAGS_WRITE)
+        strcat(buffer, "DBCOLUMNFLAGS_WRITE | ");
+    if(flags & DBCOLUMNFLAGS_WRITEUNKNOWN)
+        strcat(buffer, "DBCOLUMNFLAGS_WRITEUNKNOWN | ");
+    if(flags & DBCOLUMNFLAGS_ISFIXEDLENGTH)
+        strcat(buffer, "DBCOLUMNFLAGS_ISFIXEDLENGTH | ");
+    if(flags & DBCOLUMNFLAGS_ISNULLABLE)
+        strcat(buffer, "DBCOLUMNFLAGS_ISNULLABLE | ");
+    if(flags & DBCOLUMNFLAGS_MAYBENULL)
+        strcat(buffer, "DBCOLUMNFLAGS_MAYBENULL | ");
+    if(flags & DBCOLUMNFLAGS_ISLONG)
+        strcat(buffer, "DBCOLUMNFLAGS_ISLONG | ");
+    if(flags & DBCOLUMNFLAGS_ISROWID)
+        strcat(buffer, "DBCOLUMNFLAGS_ISROWID | ");
+    if(flags & DBCOLUMNFLAGS_ISROWVER)
+        strcat(buffer, "DBCOLUMNFLAGS_ISROWVER | ");
+    if(flags & DBCOLUMNFLAGS_CACHEDEFERRED)
+        strcat(buffer, "DBCOLUMNFLAGS_CACHEDEFERRED");
+
+    return buffer;
+}
+
 static void test_msdasql(void)
 {
     HRESULT hr;
@@ -388,6 +419,11 @@ static void test_command_text(IUnknown *cmd)
     HRESULT hr;
     OLECHAR *str;
     GUID dialect;
+    IColumnsInfo *colinfo;
+    ICommandPrepare *commandprepare;
+    DBORDINAL columns;
+    DBCOLUMNINFO *dbcolumninfo;
+    OLECHAR *buffer;
 
     hr = IUnknown_QueryInterface(cmd, &IID_ICommandText, (void**)&comand_text);
     ok(hr == S_OK, "got 0x%08x\n", hr);
@@ -395,6 +431,12 @@ static void test_command_text(IUnknown *cmd)
     hr = ICommandText_GetCommandText(comand_text, &dialect, &str);
     ok(hr == DB_E_NOCOMMAND, "got 0x%08x\n", hr);
 
+    hr = IUnknown_QueryInterface(cmd, &IID_IColumnsInfo, (void**)&colinfo);
+    ok(hr == S_OK, "got 0x%08x\n", hr);
+
+    hr = IColumnsInfo_GetColumnInfo(colinfo, &columns, &dbcolumninfo, &buffer);
+    ok(hr == DB_E_NOCOMMAND, "got 0x%08x\n", hr);
+
 if (0)
 {
     /* Crashes under windows */
@@ -411,7 +453,6 @@ if (0)
     hr = ICommandText_SetCommandText(comand_text, &DBGUID_DEFAULT, L"select * from testing");
     ok(hr == S_OK, "got 0x%08x\n", hr);
 
-
     hr = ICommandText_GetCommandText(comand_text, NULL, &str);
     ok(hr == S_OK, "got 0x%08x\n", hr);
     ok (!lstrcmpW(L"select * from testing", str), "got %s\n", debugstr_w(str));
@@ -431,6 +472,26 @@ if (0)
     ok (!lstrcmpW(L"select * from testing", str), "got %s\n", debugstr_w(str));
     HeapFree(GetProcessHeap(), 0, str);
 
+    hr = IColumnsInfo_GetColumnInfo(colinfo, &columns, &dbcolumninfo, &buffer);
+    ok(hr == DB_E_NOTPREPARED, "got 0x%08x\n", hr);
+
+    hr = ICommandText_QueryInterface(comand_text, &IID_ICommandPrepare, (void**)&commandprepare);
+    ok(hr == S_OK, "got 0x%08x\n", hr);
+
+    ok (0, "*************\n");
+    hr = ICommandPrepare_Prepare(commandprepare, 1);
+    ok(hr == S_OK, "got 0x%08x\n", hr);
+    ok (0, " *************\n");
+
+    hr = IColumnsInfo_GetColumnInfo(colinfo, &columns, &dbcolumninfo, &buffer);
+    ok(hr == S_OK, "got 0x%08x\n", hr);
+    ok(columns == 3, "got %d\n", columns);
+
+    CoTaskMemFree(dbcolumninfo);
+    CoTaskMemFree(buffer);
+
+    ICommandPrepare_Release(commandprepare);
+    IColumnsInfo_Release(colinfo);
     ICommandText_Release(comand_text);
 }
 
@@ -450,37 +511,6 @@ static void test_command_dbsession(IUnknown *cmd, IUnknown *session)
     ICommandText_Release(comand_text);
 }
 
-static char *dump_flags(DWORD flags)
-{
-    static char buffer[2048];
-
-    memset(buffer, 0, sizeof(buffer));
-    if(flags & DBCOLUMNFLAGS_ISBOOKMARK)
-        strcat(buffer, "DBCOLUMNFLAGS_ISBOOKMARK | ");
-    if(flags & DBCOLUMNFLAGS_MAYDEFER)
-        strcat(buffer, "DBCOLUMNFLAGS_MAYDEFER | ");
-    if(flags & DBCOLUMNFLAGS_WRITE)
-        strcat(buffer, "DBCOLUMNFLAGS_WRITE | ");
-    if(flags & DBCOLUMNFLAGS_WRITEUNKNOWN)
-        strcat(buffer, "DBCOLUMNFLAGS_WRITEUNKNOWN | ");
-    if(flags & DBCOLUMNFLAGS_ISFIXEDLENGTH)
-        strcat(buffer, "DBCOLUMNFLAGS_ISFIXEDLENGTH | ");
-    if(flags & DBCOLUMNFLAGS_ISNULLABLE)
-        strcat(buffer, "DBCOLUMNFLAGS_ISNULLABLE | ");
-    if(flags & DBCOLUMNFLAGS_MAYBENULL)
-        strcat(buffer, "DBCOLUMNFLAGS_MAYBENULL | ");
-    if(flags & DBCOLUMNFLAGS_ISLONG)
-        strcat(buffer, "DBCOLUMNFLAGS_ISLONG | ");
-    if(flags & DBCOLUMNFLAGS_ISROWID)
-        strcat(buffer, "DBCOLUMNFLAGS_ISROWID | ");
-    if(flags & DBCOLUMNFLAGS_ISROWVER)
-        strcat(buffer, "DBCOLUMNFLAGS_ISROWVER | ");
-    if(flags & DBCOLUMNFLAGS_CACHEDEFERRED)
-        strcat(buffer, "DBCOLUMNFLAGS_CACHEDEFERRED");
-
-    return buffer;
-}
-
 static void test_rowset_interfaces(IRowset *rowset, ICommandText *commandtext)
 {
     IRowsetInfo *info;
@@ -509,7 +539,7 @@ static void test_rowset_interfaces(IRowset *rowset, ICommandText *commandtext)
     hr = IRowset_QueryInterface(rowset, &IID_IColumnsInfo, (void**)&col_info);
     ok(hr == S_OK, "got 0x%08x\n", hr);
 
-    if (hr == S_OK)
+    if (hr == S_OK && FALSE)
     {
         DBORDINAL columns;
         DBCOLUMNINFO *dbcolumninfo;
-- 
2.33.0

